cmd/kubectl/kubectl.go

```diff
@@ -64,65 +67,105 @@ func (h *kubectlPluginHandler) Execute(executablePath string, cmdArgs, environme
-	var idx int
-	for i, arg := range os.Args {
-		if arg == "kubectl" || arg == "kc" {
-			idx = i
-			break
-		}
-	}
-
-	args := kubectl.KubectlOptions{
+	// Create a new kubectl command without a plugin handler.
+	kubectlCmd := kubectl.NewKubectlCommand(kubectl.KubectlOptions{
 		IOStreams: genericclioptions.IOStreams{
 			In:     os.Stdin,
 			Out:    os.Stdout,
 			ErrOut: os.Stderr,
 		},
-		Arguments: os.Args[idx:],
-		PluginHandler: &kubectlPluginHandler{
-			DefaultPluginHandler: kubectl.DefaultPluginHandler{
-				ValidPrefixes: plugin.ValidPluginFilenamePrefixes,
-			},
-		},
-	}
+	})
+	kubectlCmd.Aliases = []string{"kc"}
 
-	cmd := kubectl.NewDefaultKubectlCommandWithArgs(args)
-	cmd.Aliases = []string{"kc"}
+	// Add some additional kubectl flags:
+	persistentFlags := kubectlCmd.PersistentFlags()
+	logs.AddFlags(persistentFlags)                         // This is done by k8s.io/component-base/cli
+	persistentFlags.AddFlagSet(config.GetKubeCtlFlagSet()) // This is k0s specific
 
-	// Get handle on the original kubectl prerun so we can call it later
-	originalPreRunE := cmd.PersistentPreRunE
-	cmd.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
-		if err := fallbackToK0sKubeconfig(cmd); err != nil {
+	hookKubectlPluginHandler(kubectlCmd)
+	patchPluginListSubcommand(kubectlCmd)
+
+	return kubectlCmd
+}
+
+// hookKubectlPluginHandler patches the kubectl command in a way that it will
+// execute kubectl's plugin handler before actually executing the command.
+func hookKubectlPluginHandler(kubectlCmd *cobra.Command) {
+	// Intercept kubectl's flag error func, so that kubectl plugins may be
+	// handled properly, e.g. so that `k0s kc foo --bar` works as expected when
+	// there's a `kubectl-foo` plugin installed.
+	originalFlagErrFunc := kubectlCmd.FlagErrorFunc()
+	kubectlCmd.SetFlagErrorFunc(func(cmd *cobra.Command, err error) error {
+		handleKubectlPlugins(kubectlCmd)
+		return originalFlagErrFunc(cmd, err)
+	})
+
+	// Intercept kubectl's PreRunE, so that generic k0s flags are honored and
+	// kubectl plugins may be handled properly, e.g. so that `k0s kc foo bar`
+	// works as expected when there's a `kubectl-foo` plugin installed.
+	originalPreRunE := kubectlCmd.PersistentPreRunE
+	kubectlCmd.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
+		handleKubectlPlugins(kubectlCmd)
+
+		// The basic kubectl command will never accept any arguments. This will
+		// be handled more or less automatically by Cobra when used as a root
+		// command. When run as a subcommand, Cobra will instead pass all of the
+		// arguments to the command's run methods, which will trigger some
+		// unexpected output. Address this by specifying NoArgs for the kubectl
+		// command.
+		if cmd == kubectlCmd {
+			if err := cobra.NoArgs(cmd, args); err != nil {
+				return err
+			}
+		}
+
+		// In vanilla kubectl, log initialization and flushing is handled by
+		// k8s.io/component-base/cli. But k0s doesn't use it, so it needs to
```

Comment:
Side note: I was totally sure we use component-base/cli, I wonder where does this false memory come from

Reply:
I digged a bit in the GitHub history when I wrote that comment and found a superseeded PR of yours. That might be the reason:

* #1408